### PR TITLE
Generalize handling of filters and pagination with `useFilters` hook

### LINKED ISSUE #3955: Facilities, Sample Tests, Shifting and Users pagination is not using query parameters
**Describe the bug**
Facilities pagination is not using query parameters like other pages such as the patients page.

**To Reproduce**
Steps to reproduce the behaviour:

1. Go to the facilities page
2. Go to page 2 using the pagination below
3. Go back to page 1 by pressing the browser's back button

**Expected behaviour**
Land back on page 1, instead of the page before reaching the facilities page.

**Screenshots**

https://user-images.githubusercontent.com/25143503/200097331-6d97df27-c871-43ba-901a-7bb19a030309.mp4

**Tasks**
- [ ] Pagination component to be responsible for setting query parameters.
- [ ] Update the Facilities page with the updated pagination component.
- [ ] Update the Assets page with the updated pagination component.
- [ ] Update the Sample Tests page with the updated pagination component.
- [ ] Update the Shifting page with the updated pagination component.
- [ ] Update the Users page with the updated pagination component.

### LINKED ISSUE #4015: Fix issues in #4009
The following issues would be present when (#3965 #4004 #3920) are merged (draft #4009 for testing).

- #3965 
- #4004
- #3920

**Describe the bug**
- [ ] Removing the search filter tag works, but the search filter comes back after a delay.
- [ ] #3965 introduces query params-based pagination, however, #4004 still relies on the old method to solve the issue.

**To Reproduce**

(already described in the bug)

**Expected behavior**
- [ ] Removing the search filter tag, should remove it and not come back after a delay.
- [ ] Use pagination query param to reset the search filter

**Screenshots**

image

image

**Desktop (please complete the following information):**

**Smartphone (please complete the following information):**

**Additional context**

### PR DESCRIPTION
**Pull request by rithviknishad**
- **This PR is stacked on top of #3965** 

## Proposed Changes

- [x] De-duplicates redefinition of the `badge` component by making a generic component: **`FilterBadge`**.
- [x] De-duplicates redefinition of the **`applyFilter`** method.
- [x] De-duplicates redefinition of the **`removeFilter`** method.
- [x] De-duplicates redefinition of the **`handlePagination`** method.
- [x] De-duplicates redefinition of the **`updateQuery`** method.
- [x] De-duplicates recomposing of the **`Pagination`** component.
- [x] Implements **`useFilters`** hook.
- [x] Generalizes and migrates handling of filters and pagination with new hook for:
  - [x] Facilties
  - [x] Patients
  - [x] Assets
  - [x] Sample test
  - [x] Shifting
  - [x] Resource
  - [x] External results
  - [x] Users
- [x] Fixes #4015
- [x] Inherently closes #3955 via base PR #3965

@coronasafe/care-fe-code-reviewers @coronasafe/code-reviewers

## Merge Checklist

- [x] Add specs that demonstrate bug / test a new feature.
- [x] Update [product documentation](https://docs.coronasafe.network/coronasafe-care-documentation/architecture/architecture-and-layering-of-care).
- [x] Ensure that UI text is kept in I18n files.
- [x] Prep screenshot or demo video for changelog entry, and attach it to issue.
- [ ] Request for Peer Reviews
- [ ] Completion of QA

### CODE REVIEW

**Comment by nihal467** [comment]
@rithviknishad 
image

- [x] KASP embelled filters are not working

image

- [x] responsive view of the patient tab, after applying filters breaking a lot, when compared to current staging.

image

- [x] in normal view also, the UI in the patient tab is breaking

image

- [x] unable to close the district filter using badge

**Comment by rithviknishad** [comment]
image

**Comment by nihal467** [comment]
LGTM, @khavinshankar review it carefully while merging

File: src/Components/Resource/BadgesList.tsx [hunk 1]
```diff
@@ -1,152 +1,67 @@
...
+      badges={({ badge, value, boolean, dateRange }: any) => [
+        badge("Ordering", "ordering"),
+        badge("Status", "status"),
+        boolean("Emergency", "emergency", {
+          trueValue: "yes",
+          falseValue: "no",
+        }),
+        ...dateRange("Modified", "modified_date"),
+        ...dateRange("Created", "created_date"),
+        value("Origin facility", "origin_facility", orginFacilityName),
```
Comment (khavinshankar) [inline src/Components/Resource/BadgesList.tsx]:
```suggestion
        value("Origin facility", "orgin_facility", orginFacilityName),
```

File: src/Components/Shifting/BadgesList.tsx [hunk 2]
```diff
@@ -1,247 +1,96 @@
...
+  trueValue: "yes",
+  falseValue: "no",
+};
 
+export default function BadgesList(props: any) {
+  const { qParams, FilterBadges } = props;
   const [assignedUsername, setAssignedUsername] = useState("");
   const [assignedFacilityName, setAssignedFacilityName] = useState("");
-  const [orginFacilityName, setOrginFacilityName] = useState("");
+  const [originFacilityName, setOirginFacilityName] = useState("");
```
Comment (khavinshankar) [inline src/Components/Shifting/BadgesList.tsx]:
```suggestion
  const [originFacilityName, setOriginFacilityName] = useState("");
```

File: src/Components/Shifting/BadgesList.tsx [hunk 3]
```diff
@@ -1,247 +1,96 @@
...
-
-        setOrginFacilityName(res?.data?.name);
-      } else {
-        setOrginFacilityName("");
-      }
+      if (!qParams.orgin_facility) return setOirginFacilityName("");
+      const res = await dispatch(
+        getAnyFacility(qParams.orgin_facility, "orgin_facility")
+      );
+      setOirginFacilityName(res?.data?.name);
```
Comment (khavinshankar) [inline src/Components/Shifting/BadgesList.tsx]:
```suggestion
      setOriginFacilityName(res?.data?.name);
```

File: src/Components/Shifting/BadgesList.tsx [hunk 4]
```diff
@@ -1,247 +1,96 @@
...
+        boolean("Up shift", "is_up_shift", booleanFilterOptions),
+        boolean("Antenatal", "is_antenatal", booleanFilterOptions),
+        phoneNumber("Phone no.", "patient_phone_number"),
+        badge("Patient name", "patient_name"),
+        ...dateRange("Created", "created_date"),
+        badge("Disease status", "disease_status"),
+        badge("Breathlessness level", "breathlessness_level"),
+        value("Assigned to", "assigned_to", assignedUsername),
+        value("Facility assigned", "assigned_facility", assignedFacilityName),
+        value("Origin facility", "origin_facility", originFacilityName),
```
Comment (khavinshankar) [inline src/Components/Shifting/BadgesList.tsx]:
```suggestion
        value("Origin facility", "orgin_facility", originFacilityName),
```

File: src/Components/ExternalResult/ResultList.tsx [hunk 5]
```diff
@@ -486,32 +359,17 @@ export default function ResultList() {
...
+      <SlideOver {...advancedFilter}>
         <div className="bg-white min-h-screen p-4">
-          <ListFilter
-            filter={qParams}
-            onChange={applyFilter}
-            closeFilter={() => setShowFilters(false)}
-            dataList={lsgWardData}
-            local={local}
-          />
+          <ListFilter {...advancedFilter} dataList={lsgWardData} />
```
Comment (khavinshankar) [inline src/Components/ExternalResult/ResultList.tsx:372]:
```suggestion
          <ListFilter {...advancedFilter} dataList={lsgWardData} />
```

remove local in ListFilter also

**Review by khavinshankar — CHANGES_REQUESTED** [review]
@rithviknishad 
There are a few minor changes, and

It would be better if the filters remain when we change views like LIstView to BoardView and vice versa in Resources and Shifting

File: src/Components/ExternalResult/ResultList.tsx (hunk 5 above)
Comment (khavinshankar) [reply src/Components/ExternalResult/ResultList.tsx:372]: removed local filters in ListFilter

**Review by khavinshankar — APPROVED** [review]

**Pull request merged**